usr.bin/mt/mt.1
.\" Copyright (c) 1981, 1990, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"	@(#)mt.1	8.1 (Berkeley) 6/6/93
.\" $FreeBSD: src/usr.bin/mt/mt.1,v 1.19.2.8 2002/11/08 11:35:57 joerg Exp $
.\"
.Dd June 6, 1993
.Dt MT 1
.Os
.Sh NAME
.Nm mt
.Nd magnetic tape manipulating program
.Sh SYNOPSIS
.Nm
.Op Fl f Ar tapename
.Ar command
.Op Ar count
.Sh DESCRIPTION
The
.Nm
utility is used to give commands to a magnetic tape drive.
By default
.Nm
performs the requested operation once.  Operations
may be performed multiple times by specifying
.Ar count .
Note
that
.Ar tapename
must reference a raw (not block) tape device.
.Pp
The available commands are listed below.  Only as many
characters as are required to uniquely identify a command
need be specified.
.Bl -tag -width "eof, weof"
.It Cm weof
Write
.Ar count
end-of-file marks at the current position on the tape.
.It Cm smk
Write
.Ar count
setmarks at the current position on the tape.
.It Cm fsf
Forward space
.Ar count
files.
.It Cm fsr
Forward space
.Ar count
records.
.It Cm fss
Forward space
.Ar count
setmarks.
.It Cm bsf
Backward space
.Ar count
files.
.It Cm bsr
Backward space
.Ar count
records.
.It Cm bss
Backward space
.Ar count
setmarks.
.It Cm rdhpos
Read Hardware block position.
Some drives do not support this.
The block
number reported is specific for that hardware only.
The count argument is
ignored.
.It Cm rdspos
Read SCSI logical block position.
Some drives do not support this.
The
count argument is ignored.
.It Cm sethpos
Set Hardware block position.
Some drives do not support this.
The count
argument is interpreted as a hardware block to which to position the tape.
.It Cm setspos
Set SCSI logical block position.
Some drives do not support this.
The count
argument is interpreted as a SCSI logical block to which to position the tape.
.It Cm rewind
Rewind the tape
(Count is ignored).
.It Cm offline , rewoffl
Rewind the tape and place the tape unit off-line
(Count is ignored).
.It Cm erase
Erase the tape.
A count of 0 disables long erase, which is on by default.
.It Cm retension
Re-tension the tape
(one full wind forth and back, Count is ignored).
.It Cm status
Print status information about the tape unit.
For SCSI magnetic tape devices,
the current operating modes of density, blocksize, and whether compression
is enabled is reported.
The current state of the driver (what it thinks that
it is doing with the device) is reported.
If the driver knows the relative
position from BOT (in terms of filemarks and records), it prints that.
Note
that this information is not definitive (only BOT, End of Recorded Media, and
hardware or SCSI logical block position (if the drive supports such) are
considered definitive tape positions).
.It Cm errstat
Print (and clear) error status information about this device.
For every normal
operation (e.g., a read or a write) and every control operation (e.g,, a
rewind), the driver stores up the last command executed and its associated
status and any residual counts (if any). This command retrieves and prints this
information.
If possible, this also clears any latched error information.
.It Cm blocksize
Set the block size for the tape unit.  Zero means variable-length
blocks.
.It Cm density
Set the density for the tape unit.  For the density codes, see below.
The density value could be given either numerically, or as a string,
corresponding to the
.Dq Reference
field.  If the string is abbreviated, it will be resolved in the order
shown in the table, and the first matching entry will be used.  If the
given string and the resulting canonical density name do not match
exactly, an informational message is printed about what the given
string has been taken for.
.It Cm geteotmodel
Fetch and print out the current EOT filemark model.
The model states how
many filemarks will be written at close if a tape was being written.
.It Cm seteotmodel
Set (from the
.Ar count
argument)
and print out the current and EOT filemark model.
Typically this will be
.Ar 2
filemarks, but some devices (typically QIC cartridge drives) can
only write
.Ar 1
filemark.
Currently you can only choose a value of
.Ar 1
or
.Ar 2 .
.It Cm eom
Forward space to end of recorded medium
(Count is ignored).
.It Cm eod
Forward space to end of data, identical to
.Cm eom .
.It Cm comp
Set compression mode.
There are currently several possible values for the compression mode:
.Pp
.Bl -tag -width 9n -compact
.It off
Turn compression off.
.It on
Turn compression on.
.It none
Same as
.Ar off .
.It enable
Same as
.Ar on .
.It IDRC
IBM Improved Data Recording Capability compression (0x10).
.It DCLZ
DCLZ compression algorithm (0x20).
.El
.Pp
In addition to the above recognized compression keywords, the user can
supply a numeric compression algorithm for the tape drive to use.  In most
cases, simply turning the compression
.Sq on
will have the desired effect of enabling the default compression algorithm
supported by the drive.  If this is not the case (see the
.Cm status
display to see which compression algorithm is currently in use), the user
can manually specify one of the supported compression keywords (above), or
supply a numeric compression value.
.El
.Pp
If a tape name is not specified, and the environment variable
.Ev TAPE
does not exist;
.Nm
uses the device
.Pa /dev/nsa0 .
.Pp
The
.Nm
utility returns a 0 exit status when the operation(s) were successful,
1 if the command was unrecognized, and 2 if an operation failed.
.Pp
The following density table was taken from the
.Sq Historical sequential access density codes
table (A-1) in Revision 11 of the SCSI-3 Stream Device Commands (SSC)
working draft, dated November 11, 1997.
.Pp
The different density codes are as follows:
.Pp
.Dl "0x0	default for device"
.Dl "0xE	reserved for ECMA"
.Bd -literal -offset 3n
Value  Width        Tracks    Density         Code Type Reference   Note
        mm    in              bpmm       bpi
0x01   12.7  (0.5)    9         32     (800)  NRZI  R   X3.22-1983   2
0x02   12.7  (0.5)    9         63   (1,600)  PE    R   X3.39-1986   2
0x03   12.7  (0.5)    9        246   (6,250)  GCR   R   X3.54-1986   2
0x05    6.3  (0.25)  4/9       315   (8,000)  GCR   C   X3.136-1986  1
0x06   12.7  (0.5)    9        126   (3,200)  PE    R   X3.157-1987  2
0x07    6.3  (0.25)   4        252   (6,400)  IMFM  C   X3.116-1986  1
0x08    3.81 (0.15)   4        315   (8,000)  GCR   CS  X3.158-1987  1
0x09   12.7  (0.5)   18      1,491  (37,871)  GCR   C   X3.180       2
0x0A   12.7  (0.5)   22        262   (6,667)  MFM   C   X3B5/86-199  1
0x0B    6.3  (0.25)   4         63   (1,600)  PE    C   X3.56-1986   1
0x0C   12.7  (0.5)   24        500  (12,690)  GCR   C   HI-TC1       1,6
0x0D   12.7  (0.5)   24        999  (25,380)  GCR   C   HI-TC2       1,6
0x0F    6.3  (0.25)  15        394  (10,000)  GCR   C   QIC-120      1,6
0x10    6.3  (0.25)  18        394  (10,000)  GCR   C   QIC-150      1,6
0x11    6.3  (0.25)  26        630  (16,000)  GCR   C   QIC-320      1,6
0x12    6.3  (0.25)  30      2,034  (51,667)  RLL   C   QIC-1350     1,6
0x13    3.81 (0.15)   1      2,400  (61,000)  DDS   CS  X3B5/88-185A 5
0x14    8.0  (0.315)  1      1,703  (43,245)  RLL   CS  X3.202-1991  5
0x15    8.0  (0.315)  1      1,789  (45,434)  RLL   CS  ECMA TC17    5
0x16   12.7  (0.5)   48        394  (10,000)  MFM   C   X3.193-1990  1
0x17   12.7  (0.5)   48      1,673  (42,500)  MFM   C   X3B5/91-174  1
0x18   12.7  (0.5)  112      1,673  (42,500)  MFM   C   X3B5/92-50   1
0x19   12.7  (0.5)  128      2,460  (62,500)  RLL   C   DLTapeIII    6,7
0x1A   12.7  (0.5)  128      3,214  (81,633)  RLL   C   DLTapeIV(20) 6,7
0x1B   12.7  (0.5)  208      3,383  (85,937)  RLL   C   DLTapeIV(35) 6,7
0x1C    6.3  (0.25)  34      1,654  (42,000)  MFM   C   QIC-385M     1,6
0x1D    6.3  (0.25)  32      1,512  (38,400)  GCR   C   QIC-410M     1,6
0x1E    6.3  (0.25)  30      1,385  (36,000)  GCR   C   QIC-1000C    1,6
0x1F    6.3  (0.25)  30      2,666  (67,733)  RLL   C   QIC-2100C    1,6

usr.bin/mt/mt.c
/*
 * Copyright (c) 1980, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 * @(#) Copyright (c) 1980, 1993 The Regents of the University of California.  All rights reserved.
 * @(#)mt.c	8.2 (Berkeley) 5/4/95
 * $FreeBSD: src/usr.bin/mt/mt.c,v 1.26.2.3 2002/11/08 11:35:57 joerg Exp $
 */

/*
 * mt --
 *   magnetic tape manipulation program
 */
#include <sys/types.h>
#include <sys/ioctl.h>
#include <sys/mtio.h>

#include <ctype.h>
#include <err.h>
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

/* the appropriate sections of <sys/mtio.h> are also #ifdef'd for FreeBSD */
#if defined(__DragonFly__)
/* c_flags */
#define NEED_2ARGS	0x01
#define ZERO_ALLOWED	0x02
#define IS_DENSITY	0x04
#define DISABLE_THIS	0x08
#define IS_COMP		0x10
#endif /* defined(__DragonFly__) */

#ifndef TRUE
#define TRUE 1
#endif
#ifndef FALSE
#define FALSE 0
#endif

static struct commands {
	char *c_name;
	int c_code;
	int c_ronly;
#if defined(__DragonFly__)
	int c_flags;
#endif /* defined(__DragonFly__) */
} com[] = {
	{ "bsf",	MTBSF,	1 },
	{ "bsr",	MTBSR,	1 },
#if defined(__DragonFly__)
	/* XXX FreeBSD considered "eof" dangerous, since it's being
	   confused with "eom" (and is an alias for "weof" anyway) */
	{ "eof",	MTWEOF, 0, DISABLE_THIS },
#else
	{ "eof",	MTWEOF, 0 },
#endif
	{ "fsf",	MTFSF,	1 },
	{ "fsr",	MTFSR,	1 },
	{ "offline",	MTOFFL,	1 },
	{ "rewind",	MTREW,	1 },
	{ "rewoffl",	MTOFFL,	1 },
	{ "status",	MTNOP,	1 },
#if defined(__DragonFly__)
	{ "weof",	MTWEOF,	0, ZERO_ALLOWED },
#else
	{ "weof",	MTWEOF,	0 },
#endif
#if defined(__DragonFly__)
	{ "erase",	MTERASE, 0, ZERO_ALLOWED},
	{ "blocksize",	MTSETBSIZ, 0, NEED_2ARGS|ZERO_ALLOWED },
	{ "density",	MTSETDNSTY, 0, NEED_2ARGS|ZERO_ALLOWED|IS_DENSITY },
	{ "eom",	MTEOD, 1 },
	{ "eod",	MTEOD, 1 },
	{ "smk",	MTWSS, 0 },
	{ "wss",	MTWSS, 0 },
	{ "fss",	MTFSS, 1 },
	{ "bss",	MTBSS, 1 },
	{ "comp",	MTCOMP, 0, NEED_2ARGS|ZERO_ALLOWED|IS_COMP },
	{ "retension",	MTRETENS, 1 },
	{ "rdhpos",     MTIOCRDHPOS,  0 },
	{ "rdspos",     MTIOCRDSPOS,  0 },
	{ "sethpos",    MTIOCHLOCATE, 0, NEED_2ARGS|ZERO_ALLOWED },
	{ "setspos",    MTIOCSLOCATE, 0, NEED_2ARGS|ZERO_ALLOWED },
	{ "errstat",	MTIOCERRSTAT, 0 },
	{ "setmodel",	MTIOCSETEOTMODEL, 0, NEED_2ARGS|ZERO_ALLOWED },
	{ "seteotmodel",	MTIOCSETEOTMODEL, 0, NEED_2ARGS|ZERO_ALLOWED },
	{ "getmodel",	MTIOCGETEOTMODEL },
	{ "geteotmodel",	MTIOCGETEOTMODEL },
#endif /* defined(__DragonFly__) */
	{ NULL }
};

static void printreg(char *, u_int, char *);
static void status(struct mtget *);
static void usage(void) __dead2;
#if defined (__DragonFly__)
static void st_status(struct mtget *);
static int stringtodens(const char *s);
static const char *denstostring(int d);
static int denstobp(int d, int bpi);
static u_int32_t stringtocomp(const char *s);
static const char *comptostring(u_int32_t comp);
static void warn_eof(void) __dead2;
#endif /* defined (__DragonFly__) */

int
main(int argc, char **argv)
{
	struct commands *comp;
	struct mtget mt_status;
	struct mtop mt_com;
	int ch, len, mtfd;
	char *p, *tape;

	if ((tape = getenv("TAPE")) == NULL)
		tape = DEFTAPE;

	while ((ch = getopt(argc, argv, "f:t:")) != -1)
		switch(ch) {
		case 'f':
		case 't':
			tape = optarg;
			break;
		case '?':
		default:
			usage();
		}
	argc -= optind;
	argv += optind;

	if (argc < 1 || argc > 2)
		usage();

	len = strlen(p = *argv++);
	for (comp = com;; comp++) {
		if (comp->c_name == NULL)
			errx(1, "%s: unknown command", p);
		if (strncmp(p, comp->c_name, len) == 0)
			break;
	}
#if defined(__DragonFly__)
	if((comp->c_flags & NEED_2ARGS) && argc != 2)
		usage();
	if(comp->c_flags & DISABLE_THIS) {
		warn_eof();
	}
#endif /* defined(__DragonFly__) */
	if ((mtfd = open(tape, comp->c_ronly ? O_RDONLY : O_RDWR)) < 0)
		err(1, "%s", tape);
	if (comp->c_code != MTNOP) {
		mt_com.mt_op = comp->c_code;
		if (*argv) {
#if defined (__DragonFly__)
			if (!isdigit(**argv) &&
			    (comp->c_flags & IS_DENSITY)) {
				const char *dcanon;
				mt_com.mt_count = stringtodens(*argv);
				if (mt_com.mt_count == 0)
					errx(1, "%s: unknown density", *argv);
				dcanon = denstostring(mt_com.mt_count);
				if (strcmp(dcanon, *argv) != 0)
					printf(
					"Using \"%s\" as an alias for %s\n",
					       *argv, dcanon);
				p = "";
			} else if (!isdigit(**argv) &&
				   (comp->c_flags & IS_COMP)) {

				mt_com.mt_count = stringtocomp(*argv);
				if ((u_int32_t)mt_com.mt_count == 0xf0f0f0f0)
					errx(1, "%s: unknown compression",
					     *argv);
				p = "";
			} else
				/* allow for hex numbers; useful for density */
				mt_com.mt_count = strtol(*argv, &p, 0);
#else
			mt_com.mt_count = strtol(*argv, &p, 10);
#endif /* defined(__DragonFly__) */
			if ((mt_com.mt_count <=
#if defined (__DragonFly__)
			    ((comp->c_flags & ZERO_ALLOWED)? -1: 0)
			    && ((comp->c_flags & IS_COMP) == 0)
#else
			    0
#endif /* defined (__DragonFly__) */
			    ) || *p)
				errx(1, "%s: illegal count", *argv);
		}
		else
			mt_com.mt_count = 1;
#if	defined(__DragonFly__)
		switch (comp->c_code) {
		case MTIOCERRSTAT:
		{
			int i;
			union mterrstat umn;
			struct scsi_tape_errors *s = &umn.scsi_errstat;

			if (ioctl(mtfd, comp->c_code, (caddr_t)&umn) < 0)
				err(2, "%s", tape);
			(void)printf("Last I/O Residual: %u\n", s->io_resid);
			(void)printf(" Last I/O Command:");
			for (i = 0; i < sizeof (s->io_cdb); i++)
				(void)printf(" %02X", s->io_cdb[i]);
			(void)printf("\n");
			(void)printf("   Last I/O Sense:\n\n\t");
			for (i = 0; i < sizeof (s->io_sense); i++) {
				(void)printf(" %02X", s->io_sense[i]);
				if (((i + 1) & 0xf) == 0) {
					(void)printf("\n\t");
				}
			}
			(void)printf("\n");
			(void)printf("Last Control Residual: %u\n",
			    s->ctl_resid);
			(void)printf(" Last Control Command:");
			for (i = 0; i < sizeof (s->ctl_cdb); i++)
				(void)printf(" %02X", s->ctl_cdb[i]);
			(void)printf("\n");
			(void)printf("   Last Control Sense:\n\n\t");
			for (i = 0; i < sizeof (s->ctl_sense); i++) {
				(void)printf(" %02X", s->ctl_sense[i]);
				if (((i + 1) & 0xf) == 0) {
					(void)printf("\n\t");
				}
			}
			(void)printf("\n\n");
			exit(0);
			/* NOTREACHED */
		}
		case MTIOCRDHPOS:
		case MTIOCRDSPOS:
		{
			u_int32_t block;
			if (ioctl(mtfd, comp->c_code, (caddr_t)&block) < 0)
				err(2, "%s", tape);
			(void)printf("%s: %s block location %u\n", tape,
			    (comp->c_code == MTIOCRDHPOS)? "hardware" :
			    "logical", block);
			exit(0);
			/* NOTREACHED */
		}
		case MTIOCSLOCATE:
		case MTIOCHLOCATE:
		{
			u_int32_t block = (u_int32_t)mt_com.mt_count;
			if (ioctl(mtfd, comp->c_code, (caddr_t)&block) < 0)
				err(2, "%s", tape);
			exit(0);
			/* NOTREACHED */
		}
		case MTIOCGETEOTMODEL:
		{
			u_int32_t om;
			if (ioctl(mtfd, MTIOCGETEOTMODEL, (caddr_t)&om) < 0)
				err(2, "%s", tape);
			(void)printf("%s: the model is %u filemar%s at EOT\n",
			    tape, om, (om > 1)? "ks" : "k");
			exit(0);
			/* NOTREACHED */
		}
		case MTIOCSETEOTMODEL:
		{
			u_int32_t om, nm = (u_int32_t)mt_com.mt_count;
			if (ioctl(mtfd, MTIOCGETEOTMODEL, (caddr_t)&om) < 0)
				err(2, "%s", tape);
			if (ioctl(mtfd, comp->c_code, (caddr_t)&nm) < 0)
				err(2, "%s", tape);
			(void)printf("%s: old model was %u filemar%s at EOT\n",
			    tape, om, (om > 1)? "ks" : "k");
			(void)printf("%s: new model  is %u filemar%s at EOT\n",
			    tape, nm, (nm > 1)? "ks" : "k");
			exit(0);
			/* NOTREACHED */
		}
		default:
			break;
		}
#endif
		if (ioctl(mtfd, MTIOCTOP, &mt_com) < 0)
			err(1, "%s: %s", tape, comp->c_name);
	} else {
		if (ioctl(mtfd, MTIOCGET, &mt_status) < 0)
			err(1, NULL);
		status(&mt_status);
	}
	exit(0);
	/* NOTREACHED */
}

#ifdef sun
#include <sundev/tmreg.h>
#include <sundev/arreg.h>
#endif

static struct tape_desc {
	short	t_type;		/* type of magtape device */
	char	*t_name;	/* printing name */
	char	*t_dsbits;	/* "drive status" register */
	char	*t_erbits;	/* "error" register */
} tapes[] = {
#ifdef sun
	{ MT_ISCPC,	"TapeMaster",	TMS_BITS,	0 },
	{ MT_ISAR,	"Archive",	ARCH_CTRL_BITS,	ARCH_BITS },
#endif
#if defined (__DragonFly__)
	/*
	 * XXX This is weird.  The st driver reports the tape drive
	 * as 0x7 (MT_ISAR - Sun/Archive compatible); the wt driver
	 * either reports MT_ISVIPER1 for an Archive tape, or 0x11
	 * (MT_ISMFOUR) for other tapes.
	 * XXX for the wt driver, rely on it behaving like a "standard"
	 * magtape driver.
	 */
	{ MT_ISAR,	"SCSI tape drive", 0,		0 },
#endif /* defined (__DragonFly__) */
	{ 0 }
};

/*
 * Interpret the status buffer returned
 */
static void
status(struct mtget *bp)
{
	struct tape_desc *mt;

	for (mt = tapes;; mt++) {
		if (mt->t_type == 0) {
			(void)printf("%d: unknown tape drive type\n",
			    bp->mt_type);
			return;
		}
		if (mt->t_type == bp->mt_type)
			break;
	}
#if defined (__DragonFly__)
	if(mt->t_type == MT_ISAR)
		st_status(bp);
	else {
#endif /* defined (__DragonFly__) */
	(void)printf("%s tape drive, residual=%d\n", mt->t_name, bp->mt_resid);
	printreg("ds", (unsigned short)bp->mt_dsreg, mt->t_dsbits);
	printreg("\ner", (unsigned short)bp->mt_erreg, mt->t_erbits);
	(void)putchar('\n');
#if defined (__DragonFly__)
	}
#endif /* defined (__DragonFly__) */
}

/*
 * Print a register a la the %b format of the kernel's printf.
 */
static void
printreg(char *s, u_int v, char *bits)
{
	int i, any = 0;
	char c;

	if (bits && *bits == 8)
		printf("%s=%o", s, v);
	else
		printf("%s=%x", s, v);
	if (!bits)
		return;
	bits++;
	if (v && bits) {
		putchar('<');
		while ((i = *bits++)) {
			if (v & (1 << (i-1))) {
				if (any)
					putchar(',');
				any = 1;
				for (; (c = *bits) > 32; bits++)
					putchar(c);
			} else
				for (; *bits > 32; bits++)
					;
		}
		putchar('>');
	}
}

static void
usage(void)
{
	(void)fprintf(stderr, "usage: mt [-f device] command [ count ]\n");
	exit(1);
}

#if defined (__DragonFly__)

static struct densities {
	int dens;
	int bpmm;
	int bpi;
	const char *name;
} dens[] = {
	/*
	 * Taken from T10 Project 997D 
	 * SCSI-3 Stream Device Commands (SSC)
	 * Revision 11, 4-Nov-97
	 */
	/*Num.  bpmm    bpi     Reference     */
	{ 0x1,	32,	800,	"X3.22-1983" },
	{ 0x2,	63,	1600,	"X3.39-1986" },
	{ 0x3,	246,	6250,	"X3.54-1986" },
	{ 0x5,	315,	8000,	"X3.136-1986" },
	{ 0x6,	126,	3200,	"X3.157-1987" },
	{ 0x7,	252,	6400,	"X3.116-1986" },
	{ 0x8,	315,	8000,	"X3.158-1987" },
	{ 0x9,	491,	37871,	"X3.180" },
	{ 0xA,	262,	6667,	"X3B5/86-199" },
	{ 0xB,	63,	1600,	"X3.56-1986" },
	{ 0xC,	500,	12690,	"HI-TC1" },
	{ 0xD,	999,	25380,	"HI-TC2" },
	{ 0xF,	394,	10000,	"QIC-120" },
	{ 0x10,	394,	10000,	"QIC-150" },
	{ 0x11,	630,	16000,	"QIC-320" },
	{ 0x12,	2034,	51667,	"QIC-1350" },
	{ 0x13,	2400,	61000,	"X3B5/88-185A" },
	{ 0x14,	1703,	43245,	"X3.202-1991" },
	{ 0x15,	1789,	45434,	"ECMA TC17" },
	{ 0x16,	394,	10000,	"X3.193-1990" },
	{ 0x17,	1673,	42500,	"X3B5/91-174" },
	{ 0x18,	1673,	42500,	"X3B5/92-50" },
	{ 0x19, 2460,   62500,  "DLTapeIII" },
	{ 0x1A, 3214,   81633,  "DLTapeIV(20GB)" },
	{ 0x1B, 3383,   85937,  "DLTapeIV(35GB)" },
	{ 0x1C, 1654,	42000,	"QIC-385M" },
	{ 0x1D,	1512,	38400,	"QIC-410M" },
	{ 0x1E, 1385,	36000,	"QIC-1000C" },
	{ 0x1F,	2666,	67733,	"QIC-2100C" },
	{ 0x20, 2666,	67733,	"QIC-6GB(M)" },
	{ 0x21,	2666,	67733,	"QIC-20GB(C)" },
	{ 0x22,	1600,	40640,	"QIC-2GB(C)" },
	{ 0x23, 2666,	67733,	"QIC-875M" },
	{ 0x24,	2400,	61000,	"DDS-2" },
	{ 0x25,	3816,	97000,	"DDS-3" },
	{ 0x26,	3816,	97000,	"DDS-4" },
	{ 0x27,	3056,	77611,	"Mammoth" },
	{ 0x28,	1491,	37871,	"X3.224" },
	{ 0x41, 3868,   98250,  "DLTapeIV(40GB)" },
	{ 0x48, 5236,   133000, "SDLTapeI(110)" },
	{ 0x49, 7598,   193000, "SDLTapeI(160)" },
	{ 0, 0, 0, NULL }
};

static struct compression_types {
	u_int32_t	comp_number;
	const char 	*name;
} comp_types[] = {
	{ 0x00, "none" },
	{ 0x00, "off" },
	{ 0x10, "IDRC" },
	{ 0x20, "DCLZ" },
	{ 0xffffffff, "enable" },
	{ 0xffffffff, "on" },
	{ 0xf0f0f0f0, NULL}
};

static const char *
denstostring(int d)
{
	static char buf[20];
	struct densities *sd;

	/* densities 0 and 0x7f are handled as special cases */
	if (d == 0)
		return "default";
	if (d == 0x7f)
		return "same";
	for (sd = dens; sd->dens; sd++)
		if (sd->dens == d)
			break;
	if (sd->dens == 0)
		sprintf(buf, "0x%02x", d);
	else 
		sprintf(buf, "0x%02x:%s", d, sd->name);
	return buf;
}

/*
 * Given a specific density number, return either the bits per inch or bits
 * per millimeter for the given density.
 */
static int
denstobp(int d, int bpi)
{
	struct densities *sd;

	for (sd = dens; sd->dens; sd++)
		if (sd->dens == d)
			break;
	if (sd->dens == 0)
		return(0);
	else {
		if (bpi)
			return(sd->bpi);
		else
			return(sd->bpmm);
	}
}

static int
stringtodens(const char *s)
{
	struct densities *sd;
	size_t l = strlen(s);

	for (sd = dens; sd->dens; sd++)
		if (strncasecmp(sd->name, s, l) == 0)
			break;
	return sd->dens;
}


static const char *
getblksiz(int bs)
{
	static char buf[25];
	if (bs == 0)
		return "variable";
	else {
		sprintf(buf, "%d bytes", bs);
		return buf;
	}
}

static const char *
comptostring(u_int32_t comp)
{
	static char buf[20];
	struct compression_types *ct;

	if (comp == MT_COMP_DISABLED)
		return "disabled";
	else if (comp == MT_COMP_UNSUPP)
		return "unsupported";

	for (ct = comp_types; ct->name; ct++)
		if (ct->comp_number == comp)
			break;

	if (ct->comp_number == 0xf0f0f0f0) {
		sprintf(buf, "0x%x", comp);
		return(buf);
	} else
		return(ct->name);
}

static u_int32_t
stringtocomp(const char *s)
{
	struct compression_types *ct;
	size_t l = strlen(s);

	for (ct = comp_types; ct->name; ct++)
		if (strncasecmp(ct->name, s, l) == 0)
			break;

	return(ct->comp_number);
}

static void
st_status(struct mtget *bp)
{
	printf("Mode      Density              Blocksize      bpi      "
	       "Compression\n"
	       "Current:  %-17s    %-12s   %-7d  %s\n"
	       "---------available modes---------\n"
	       "0:        %-17s    %-12s   %-7d  %s\n"
	       "1:        %-17s    %-12s   %-7d  %s\n"
	       "2:        %-17s    %-12s   %-7d  %s\n"
	       "3:        %-17s    %-12s   %-7d  %s\n",
	       denstostring(bp->mt_density), getblksiz(bp->mt_blksiz),
	       denstobp(bp->mt_density, TRUE), comptostring(bp->mt_comp),
	       denstostring(bp->mt_density0), getblksiz(bp->mt_blksiz0),
	       denstobp(bp->mt_density0, TRUE), comptostring(bp->mt_comp0),
	       denstostring(bp->mt_density1), getblksiz(bp->mt_blksiz1),
	       denstobp(bp->mt_density1, TRUE), comptostring(bp->mt_comp1),
	       denstostring(bp->mt_density2), getblksiz(bp->mt_blksiz2),
	       denstobp(bp->mt_density2, TRUE), comptostring(bp->mt_comp2),
	       denstostring(bp->mt_density3), getblksiz(bp->mt_blksiz3),
	       denstobp(bp->mt_density3, TRUE), comptostring(bp->mt_comp3));

	if (bp->mt_dsreg != MTIO_DSREG_NIL) {
		auto char foo[32];
		const char *sfmt = "Current Driver State: %s.\n";
		printf("---------------------------------\n");
		switch (bp->mt_dsreg) {
		case MTIO_DSREG_REST:
			printf(sfmt, "at rest");      
			break;
		case MTIO_DSREG_RBSY:    
			printf(sfmt, "Communicating with drive");
			break;
		case MTIO_DSREG_WR:
			printf(sfmt, "Writing");
			break;
		case MTIO_DSREG_FMK:
			printf(sfmt, "Writing Filemarks");
			break;
		case MTIO_DSREG_ZER:
			printf(sfmt, "Erasing");
			break;
		case MTIO_DSREG_RD:
			printf(sfmt, "Reading");
			break;
		case MTIO_DSREG_FWD:
			printf(sfmt, "Spacing Forward");
			break;
		case MTIO_DSREG_REV:     
			printf(sfmt, "Spacing Reverse");
			break;
		case MTIO_DSREG_POS:
			printf(sfmt,
			    "Hardware Positioning (direction unknown)");
			break;
		case MTIO_DSREG_REW:
			printf(sfmt, "Rewinding");
			break;
		case MTIO_DSREG_TEN:
			printf(sfmt, "Retensioning");
			break;
		case MTIO_DSREG_UNL:
			printf(sfmt, "Unloading");
			break;
		case MTIO_DSREG_LD:
			printf(sfmt, "Loading");
			break;
		default:
			(void) sprintf(foo, "Unknown state 0x%x", bp->mt_dsreg);
			printf(sfmt, foo);
			break;
		}
	}
	if (bp->mt_resid == 0 && bp->mt_fileno == (daddr_t) -1 &&
	    bp->mt_blkno == (daddr_t) -1)
		return;
	printf("---------------------------------\n");
	printf("File Number: %u\tRecord Number: %u\tResidual Count %d\n",
	    bp->mt_fileno, bp->mt_blkno, bp->mt_resid);
}

static void
warn_eof(void)
{
	fprintf(stderr,
		"The \"eof\" command has been disabled.\n"
		"Use \"weof\" if you really want to write end-of-file marks,\n"
		"or \"eom\" if you rather want to skip to the end of "
		"recorded medium.\n");
	exit(1);
}

#endif /* defined (__DragonFly__) */
